src/transformers/tokenization_utils.py

```diff
@@ -422,6 +422,9 @@ def _add_tokens(self, new_tokens: Union[List[str], List[AddedToken]], special_to
                 and token not in tokens_to_add
             ):
                 tokens_to_add.append(token)
+                if isinstance(new_tokens[i], AddedToken) or special_tokens:
+                    # tokens that are added using AddedToken are special tokens.
+                    self._additional_special_tokens.append(new_tokens[i])
```

Comment:
This fixes the issue of losing the information on `rstrip` and  `lstrip` for any of the added tokens, but also the decoding issue since `_decode` checks if the tokens are in the `all_special_ids`, which (after 4 call to properties) checks the `_additional_special_tokens`.